src/core/lib/channel/promise_based_filter.h

```diff
@@ -263,6 +269,92 @@ class BaseCallData : public Activity, private Wakeable {
     return p.release();
   }
 
+  class SendMessage {
+   public:
+    explicit SendMessage(BaseCallData* base)
+        : base_(base), pipe_(base->arena()) {}
+    PipeReceiver<MessageHandle>* outgoing_pipe() { return &pipe_.receiver; }
+
+    void StartOp(CapturedBatch batch);
+    void GotPipe(PipeReceiver<MessageHandle>* receiver);
+    void WakeInsideCombiner(Flusher* flusher);
+    void Done(const ServerMetadata& metadata);
+    bool HaveCapturedBatch() const { return batch_.is_captured(); }
+    bool IsIdle() const;
+
+   private:
+    enum class State : uint8_t {
+      kInitial,
```

Comment:
Please add comments for each of the state values, in particular covering when we get into that state and what possible transitions are out of the state.

Reply:
done